ESQL: Improve detection of @timestamp inside query

costin opened the pull request:
For better selectivity of canMatch, detect if the given query has any
 filters on @timestamp that can be pushed down before sending requests
 to the cluster node.
 This should help eliminate more nodes that do not match the criteria

Related #99073

costin committed:
ESQL: Improve detection of @timestamp inside query

For better selectivity of canMatch, detect if the given query has any
 filters on @timestamp that can be pushed down before sending requests
 to the cluster node.
 This should help eliminate more nodes that do not match the criteria

Related #99073

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Hi @costin, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

costin committed:
Remove String.formatted method

luigidellaquila reviewed (changes requested):
Cool! This could be a critical optimization for many typical scenarios.

The checks on the references should be narrowed to FieldAttributes only though, eg. the following test will fail

```
    public void testTimestampAsFunctionArgument() {
        var eqValue = 1234;

        var plan = plan(LoggerMessageFormat.format(null, """
             FROM test
            |WHERE to_int(to_string({})) == {}
            """, AT_TIMESTAMP, eqValue), null);

        var filter = filterForTransportNodes(plan);
        assertThat(filter, nullValue());
    }

```

costin committed:
Address feedback

costin commented:
Thanks Luigi for catching the function case. I've reused the logic from the optimization rule since it already does traversal and added the suggested test and a variation of it.
Also renamed the method names slightly to help with reviews since I've used `var` everywhere.

luigidellaquila reviewed (approved):
LGTM, thanks!

astefan reviewed (approved):
LGTM
But I am a bit  concerned because there is no test that proves the filter is actually getting pushed to Lucene in that specific form. Maybe this is not this PR's concern since the actual push is an optimization that is included in the `filter`  functionality.

I've tested this with a `filter`:

```
{
    "query": "from endgame-140 | keep @timestamp, unique_pid, unique_ppid, user_name, timestamp | where not (@timestamp < \"2000-01-01\") | where @timestamp < \"1999-01-01\"",
    "filter": {
        "bool": {
            "filter": {
                "range": {
                    "unique_pid": {
                        "gt": 0
                    }
                }
            },
            "must": {
                "range": {
                    "@timestamp": {
                        "gt": "1950-01-01"
                    }
                }
            }
        }
    }
}
```

And I could see in the logs the actual query in the form of:

```
ExchangeSinkExec[[@timestamp{f}#263, unique_pid{f}#265, unique_ppid{f}#266, user_name{f}#267, timestamp{f}#264]]
\_FragmentExec[filter={
  "bool" : {
    "must" : [
```

But if in the query above I remove the `filter` the query using only `@timestamp` is not visible at all. It would be good to see what exactly is getting pushed to Lucene for debugging purposes in case we need this aid.

costin committed:
Add logging statement

costin commented:
> It would be good to see what exactly is getting pushed to Lucene for debugging purposes in case we need this aid.

Good idea, I've added a logging statement in ComputeService with the plan and filter sent from the coordinator to the data nodes.
